#! /bin/sh -e

# DP: gcc472 fixes and workarounds

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p1 < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p1 < $0;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

Index: glibc-2.5/elf/Makefile
===================================================================
--- glibc-2.5.orig/elf/Makefile	2014-01-19 11:00:26.298571996 +0200
+++ glibc-2.5/elf/Makefile	2014-01-19 11:00:48.930572378 +0200
@@ -305,7 +305,7 @@
 		  $(LDFLAGS-rtld) -Wl,-z,defs -Wl,--verbose 2>&1 |	\
 		  LC_ALL=C \
 		  sed -e '/^=========/,/^=========/!d;/^=========/d'	\
-		      -e 's/\. = 0 + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
+		      -e 's/\. = .* + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
 		  > $@.lds
 	$(LINK.o) -nostdlib -nostartfiles -shared -o $@			\
 		  $(LDFLAGS-rtld) -Wl,-z,defs $(z-now-$(bind-now))	\
Index: glibc-2.5/intl/dcigettext.c
===================================================================
--- glibc-2.5.orig/intl/dcigettext.c	2014-01-19 11:00:38.930572211 +0200
+++ glibc-2.5/intl/dcigettext.c	2014-01-19 11:01:08.858572714 +0200
@@ -287,7 +287,7 @@
 #ifdef _LIBC
 # include "../locale/localeinfo.h"
 # define category_to_name(category) \
-  _nl_category_names.str + _nl_category_name_idxs[category]
+  (_nl_category_names.str + *(_nl_category_name_idxs + category))
 #else
 static const char *category_to_name PARAMS ((int category)) internal_function;
 #endif
Index: glibc-2.5/nscd/Makefile
===================================================================
--- glibc-2.5.orig/nscd/Makefile	2014-01-19 11:00:31.874572091 +0200
+++ glibc-2.5/nscd/Makefile	2014-01-19 11:01:01.042572582 +0200
@@ -94,7 +94,7 @@
 nscd-cflags += -fpie
 endif
 ifeq (yes,$(have-ssp))
-nscd-cflags += -fstack-protector
+nscd-cflags += -fstack-protector $(common-objpfx)$(binfmt-subdir)/ld-linux.so.3
 endif
 
 CFLAGS-nscd.c += $(nscd-cflags)
